rasa/nlu/training_data/formats/markdown.py

```diff
@@ -211,23 +211,31 @@ def dumps(self, training_data: "TrainingData") -> Text:
     def _generate_training_examples_md(self, training_data: "TrainingData") -> Text:
         """Generates markdown training examples."""
 
-        # Sort training data by intent
-        training_examples = sorted(
-            [e.as_dict_nlu() for e in training_data.training_examples],
-            key=lambda k: k[MESSAGE_INTENT_ATTRIBUTE],
-        )
+        from collections import defaultdict
+
+        training_examples = defaultdict(list)
+
+        # Sort by intent while keeping basic intent order
+        for example in [e.as_dict_nlu() for e in training_data.training_examples]:
+            intent = example[MESSAGE_INTENT_ATTRIBUTE]
+            training_examples[intent].append(example)
+
         md = ""
-        for i, example in enumerate(training_examples):
-            intent = training_examples[i - 1][MESSAGE_INTENT_ATTRIBUTE]
-            if i == 0 or intent != example[MESSAGE_INTENT_ATTRIBUTE]:
-                md += self._generate_section_header_md(
-                    INTENT,
-                    example[MESSAGE_INTENT_ATTRIBUTE],
-                    example.get(MESSAGE_RESPONSE_KEY_ATTRIBUTE, None),
-                    i != 0,
-                )
 
-            md += self._generate_item_md(self._generate_message_md(example))
+        for intent, examples in training_examples.items():
+            for i, example in enumerate(examples):
+                did_intent_change = i == 0
+                is_first_line = md == ""
+
+                if did_intent_change:
+                    md += self._generate_section_header_md(
+                        INTENT,
+                        intent,
+                        example.get(MESSAGE_RESPONSE_KEY_ATTRIBUTE, None),
```

Comment:
regarding this argument: does this actually work? e.g. should we not rather create one section per `intent/response_key`? instead of merging all possibly different response keys into one section

Reply:
@dakshvar22

Reply:
@tmbo That is what the third argument is. The `response_key` is used here - https://github.com/RasaHQ/rasa/blob/cad0fc00627b46c96ba10ba094d69ea9676f2d69/rasa/nlu/training_data/formats/markdown.py#L289 to create a `/response_key` part of `intent/response_key` if `response_key` is set for the message